atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/RetryOnSocketTimeoutExceptionProxy.java

```diff
@@ -0,0 +1,96 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.http.v2;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.reflect.AbstractInvocationHandler;
+import com.palantir.common.base.Throwables;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.net.SocketTimeoutException;
+import java.util.function.Supplier;
+
+/**
+ * This proxy exists to retry in case of a SocketTimeoutException (something that isn't retried natively by Dialogue).
+ */
+public final class RetryOnSocketTimeoutExceptionProxy<T> extends AbstractInvocationHandler {
+    private static final int MAX_NUM_RETRIES = 5;
+    private final Supplier<T> delegate;
+
+    private static int numRetries = 0;
+
+    private RetryOnSocketTimeoutExceptionProxy(Supplier<T> delegate) {
+        this.delegate = delegate;
+    }
+
+    public static <U> U newProxyInstance(Class<U> interfaceClass, Supplier<U> delegate) {
+        RetryOnSocketTimeoutExceptionProxy<U> proxy = new RetryOnSocketTimeoutExceptionProxy<>(delegate);
+
+        return (U) Proxy.newProxyInstance(interfaceClass.getClassLoader(), new Class<?>[] {interfaceClass}, proxy);
+    }
+
+    @Override
+    protected Object handleInvocation(Object proxy, Method method, Object[] args) throws Throwable {
+        ResultOrThrowable attempt = singleInvocation(method, args);
+        while ((numRetries < MAX_NUM_RETRIES) && !attempt.isSuccessful()) {
+            numRetries++;
+            Throwable cause = attempt.throwable().get();
+            ResultOrThrowable retriableExceptionCheck = isRetriable(cause);
+            if (!retriableExceptionCheck.isSuccessful()) {
+                throw retriableExceptionCheck.throwable().get();
+            }
+            attempt = singleInvocation(method, args);
+        }
+        if (attempt.isSuccessful()) {
+            return attempt.result().orElse(null);
+        }
+        throw Throwables.unwrapIfPossible(attempt.throwable().get());
+    }
+
+    private ResultOrThrowable singleInvocation(Method method, Object[] args) {
+        try {
+            return ResultOrThrowable.success(method.invoke(delegate.get(), args));
+        } catch (Throwable th) {
+            return ResultOrThrowable.failure(th);
+        }
+    }
+
+    private ResultOrThrowable isRetriable(Throwable throwable) {
+        if (!(throwable instanceof InvocationTargetException)) {
+            return ResultOrThrowable.failure(throwable);
+        }
+        InvocationTargetException exception = (InvocationTargetException) throwable;
+        Throwable cause = exception.getCause();
+        if (!isCausedBySocketTimeoutException(cause)) {
+            return ResultOrThrowable.failure(cause);
+        }
+        return ResultOrThrowable.success(null);
```

Comment:
Flip the if, and I'd add the other part in the else block.

```suggestion
        if (isCausedBySocketTimeoutException(cause)) {
            return ResultOrThrowable.success(null);
        } else {
            return ResultOrThrowable.failure(cause);
        }
```